### PR TITLE
Xnb lattice collection

### PR DESCRIPTION
**Pull request by lafourcadep**
Added custom lattice definition and cleaned all lattice generator to avoid code duplication. 
The different lattice operator are the following : 
lattice_generator.cpp                  -> baseline operator
legacy_lattice_generator.cpp     -> similar but with internal init_rcb_grid (Additional boolean)
bulk_lattice_generator.cpp         -> similar but this one initializes the domain on its own and is dedicated to 3D periodic boundary conditions! specific to MD where the entire domain is filled with a specific lattice that needs to be perfectly commensurated with the domain.

Two other PRs will be opened for exaStamp and exaDEM to which I added examples for test purposes.

### CODE REVIEW

File: src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/lattice_generator.h [hunk 1]
```diff
@@ -94,6 +97,96 @@ namespace exanb
...
+      lattice.m_structure = *structure;
+      lattice.m_size = *size;
+      if (*structure == "CUSTOM") {
+        assert( (*np).has_value() );
+        assert( (*positions).has_value() );
+        lattice.m_np = *np;
+        lattice.m_types = *types;
+        // Checking that positions are contained between 0 and 1 -> fractional coordinates
+        for (int i=0;i<(*positions).size();i++) {
+          double px = (*positions)[i].x;
```
Comment (rprat-pro) [inline src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/lattice_generator.h:114]: Add `[[maybe_unused]]` here for compiler or define this section `#ifndef NDEGUB` (avoid warning)

px, py, pz

File: src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/lattice_generator.h [hunk 2]
```diff
@@ -94,6 +97,96 @@ namespace exanb
...
+      LatticeCollection lattice;
+      lattice.m_structure = *structure;
+      lattice.m_size = *size;
+      if (*structure == "CUSTOM") {
+        assert( (*np).has_value() );
+        assert( (*positions).has_value() );
+        lattice.m_np = *np;
+        lattice.m_types = *types;
+        // Checking that positions are contained between 0 and 1 -> fractional coordinates
+        for (int i=0;i<(*positions).size();i++) {
```
Comment (rprat-pro) [inline src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/lattice_generator.h:113]: size_t i instead of i

**Review by rprat-pro — CHANGES_REQUESTED** [review]
@lafourcadep @carrardt I confirm that it works with exaDEM =)

**Review by carrardt — APPROVED** [review]
Everything's fine, thank's for the work

**Pull request merged**